File: frontend/src/services/currency-splitter.ts
import { inject, Injectable } from '@angular/core';
import { DefaultService } from '../openapi';
import { map, Observable } from 'rxjs';
import { GlobalState } from './global-state';

export type SplitResult = { [key: string]: number };

@Injectable({ providedIn: 'root' })
export class CurrencySplitter {
  private globalState = inject(GlobalState);
  private api = inject(DefaultService);
  private currencyValues = this.api.apiValuesGet();

  /**
   * Optimally splits a provided sum of money into banknotes and coins.
   * The result is an object with the required banknote/coin values as keys.
   * The associated value is the required amount of that specific banknote/coin.
   *
   * @param total The sum of money to be split.
   * @returns An Observable of an object containing the amount of all banknotes/coins the total has been split into.
   */
  public split(total: number): Observable<SplitResult> {
    return this.globalState.calculateLocally()
      ? this.splitLocally(total)
      : this.splitRemotely(total);
  }

  private splitRemotely(total: number): Observable<SplitResult> {
    return this.api.apiSplitGet(total);
  }

  private splitLocally(total: number): Observable<SplitResult> {
    return this.currencyValues.pipe(
      map((values) => {
        const result: SplitResult = {};
        let remainder: number | undefined = undefined;

        for (const value of values) {
          const dividend: number = remainder ?? total;
          const quotient = Math.floor(dividend / value);

          if (quotient > 0) {
            result[value] = quotient;
            remainder = dividend % value;
          }
        }

        return result;
      })
    );
  }
}

File: backend/src/main/java/com/abramu/jakarta/services/CurrencySplitter.java
package com.abramu.jakarta.services;

import java.math.BigDecimal;
import java.util.Map;
import java.util.HashMap;
import java.util.List;
import java.util.Optional;

import jakarta.enterprise.context.ApplicationScoped;

@ApplicationScoped
public class CurrencySplitter {
    
    /**
     * Optimally splits a provided sum of money into banknotes and coins.
     * The possible values of all banknotes/coins must be provided in descending order, e.g. "5, 2, 1".
     * The result is a Map with the required banknote/coin values as keys.
     * The associated value is the required amount of that specific banknote/coin.
     * 
     * @param total The sum of money to be split.
     * @param currencyValues All values of the banknotes/coins that the total can be split into.
     * @return A Map containing the amount of all banknotes/coins the total has been split into.
     */
    public Map<BigDecimal, Integer> split(BigDecimal total, List<BigDecimal> currencyValues) {
        final var map = new HashMap<BigDecimal, Integer>();
        Optional<BigDecimal> remainder = Optional.empty();
        
        for (var currencyValue : currencyValues) {
            final var dividend = remainder.orElse(total);
            final var result = dividend.divideAndRemainder(currencyValue);
            final var quotient = result[0].intValue();
            final var newRemainder = result[1];
            
            if (quotient > 0) {
                map.put(currencyValue, quotient);
                remainder = Optional.of(newRemainder);
            }
        }

        return map;
    }
}
